README.md
# Log Memusage
## peak_memusage
This utility is a wrapper around getrusage, which reports peak memory
use of any executable. It is OpenMP and MPI aware and tries to report
thread- and task- specific data. Of course, being OpenMP shared memory,
that report cannot really separate thread-specific memory use.

Note that the memory usage is reported only at the end of the run.

```pre
NAME
       peak_memusage - Determine the high-water mark memory usage of an application.

SYNOPSIS
       $ peak_memusage <program> --args --for-program

       $ mpiexec peak_memusage <mpi_program>

DESCRIPTION
       peak_memusage determines the high-water mark of an applciation's memory usage.

       This  utility  is  a wrapper around getrusage, which reports peak memory use of any executable. It is OpenMP and MPI aware and tries to report thread- and task- specific data. Of course, being OpenMP shared
       memory, that report cannot really separate thread-specific memory use.

       Note that the memory usage is reported only at the end of the run.

OPTIONS
       The peak_memusage does not take any options.

EXAMPLES
       $ ./peak_memusage ./use_memory
       Running: ./use_memory; sleep 1  - Please wait...
       Allocating, using and freeing 1000000 ints (3.81 MiB) in thread 0/1
       casper-login2 used memory in task 0: 30.82MiB (+23.03MiB overhead). ExitStatus: 0. Signal: 0
```

## liblog_memusage.so

```pre
NAME
       log_memusage - Logs memory usage history of an application.

SYNOPSIS
       $ gcc -o my_app my_app.c -llog_memusage
       $ ./my_app

       $ mpiexec -x LD_PRELOAD=/path/to/liblog_memusage.so <mpi_program>

DESCRIPTION
       log_memusage is a library that monitors and reports application memory usage.~

ENVIRONMENT
   General Environment Variables
       LOG_MEMUSAGE
           If enabled (1), logs memory usage vs. time to the output file "memory_usage.log".

           Default: 0

       LOG_MEMUSAGE_POLL_INTERVAL
           Floating point value specifying the polling frequency (in seconds).

           Default: 0.1

        LOG_MEMUSAGE_OUTPUT_INTERVAL
           Integer value specifying output frequency, in multiples of LOG_MEMUSAGE_POLL_INTERVAL.  For example, a value of 10 will update the memory usage output log file every 10 polling intervals.

           Default: 1

EXAMPLES
   Baseline, uninstrumented program:
       $ ./use_memory -s 50000000
       Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/1

   Using LD_PRELOAD to instrument an existing program:
       $ LD_PRELOAD=./liblog_memusage.so ./use_memory -s 50000000
       logging memory usage for process 103284 to memory_usage.log ...
       Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/1
       # (memusage) --> casper-login2 / PID 103284, peak used memory: 213 MiB

   Using LD_PRELOAD to instrument an existing OpenMP program:
       $ LD_PRELOAD=./liblog_memusage.so OMP_NUM_THREADS=6 ./use_memory_openmp -s 50000000
       logging memory usage for process 106876 to memory_usage.log ...
       Allocating, using and freeing 100000000 ints (381.47 MiB) in thread 1/6
       Allocating, using and freeing 200000000 ints (762.94 MiB) in thread 3/6
       Allocating, using and freeing 150000000 ints (572.20 MiB) in thread 2/6
       Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/6
       Allocating, using and freeing 250000000 ints (953.67 MiB) in thread 4/6
       Allocating, using and freeing 300000000 ints (1144.41 MiB) in thread 5/6
       # (memusage) --> casper-login2 / PID 106876, peak used memory: 4036 MiB

   Using pkg-config configuration to link with an application:
       compile:
       $ gcc -o use_memory ./use_memory.c $(pkg-config log_memusage --libs --cflags)

       run:
       $ ./use_memory
       logging memory usage for process 111615 to memory_usage.log ...
       Allocating, using and freeing 1000000 ints (3.81 MiB) in thread 0/1
       # (memusage) --> casper-login2 / PID 111615, peak used memory: 24 MiB
```

RELEASE_NOTES
********************************************************************
v3.0.0: January 2023

* Refactored build system to use GNU Autotools.

* Introduced liblog_memusage.so to create a time-history of memory
  usage, for both CPUs and GPUs.


********************************************************************
v1.0.2

* Installation script for https://github.com/NCAR/HPCinstall/ added

* Stale LSF-specific scripts removed


********************************************************************
v1.0.1

* This is basically the same as v1.0 with minor change in the
  names and some printed text, made public

********************************************************************

v1.0

* Linux kernel v2.6.9 fixed the outstanding issues preventing this
  tool from working on linux, so this version has been ported
  to the linux+gcc arch

* The output has been made more clear and compact

* A few bugs (mostly related to the seldom used OpenMP version) have
  been fixed

* The serial version is now able to provide MPI information without
  using any MPI call

* The report now it's printed in stderr

********************************************************************

v0.9

* returns the program exit code and the signal sent to it, if
  any


********************************************************************

v0.8

Minor changes in the output, in the help message and in the
build process

********************************************************************

v0.72
$Date$
$Author$
$Rev$

Some spelling mistakes fixed.
Name changed to job_memusage

********************************************************************

v0.71
Date: 2008-01-24 12:05:23 -0700 (Thu, 24 Jan 2008)
Author: ddvento
Rev: 25

This tool will run the given executable and at then will print out the
memory usage for it. It works for interactive, OpenMP, MPI and
hybrid executables.

Usage:
job_memusage.exe [--details] filename-to-run [arguments-to-pass]

--details = detailed view, usually not required (must placed BEFORE everything else)

Notes:
1) for OpenMP (and hybrid), the memory is shared, so only the total
amount per node is printed: no per-thread details

2) for MPI (and hybrid), if tasks (i.e. MPI rank) identification is
required, the MP_LABELIO environment variable must be set to yes

Examples:

/contrib/bin/job_memusage.exe ./hello_world.exe

/contrib/bin/job_memusage.exe --details cat *.txt

export MP_LABELIO=yes #csh users: setenv MP_LABELIO yes
mpirun.lsf /contrib/bin/job_memusage.exe ./cam < namelist


********************************************************************

v0.7
Date: 2008-01-24 (Thu, 23 Jan 2008)
Author: ddvento

Works always: interactively, MPI and OpenMP. Requires the environment
variable MP_LABELIO set to yes for MPI task identification.

Command Line Usage:
/contrib/bin/job_memusage.exe [--details] filename-to-run [arguments-to-pass]

OpenMP Usage:
As in Command Line, but the executable should be compiled with -qsmp=omp
(gothca: you will see only the total memory, which is not too bad because it's shared)

MPI Usage (for ksh):
export MP_LABELIO=yes
mpirun.lsf /contrib/bin/job_memusage.exe [--details] filename-to-run [arguments-to-pass]


MPI Usage (for csh):
setenv MP_LABELIO yes
mpirun.lsf /contrib/bin/job_memusage.exe [--details] filename-to-run [arguments-to-pass]


In all cases:

--details = detailed view, usually not required (must placed BEFORE everything else)

********************************************************************

v0.6

Works only for MPI tasks.

Usage:
mpirun.lsf /contrib/bin/job_memusage.exe [--details] filename-to-run [arguments-to-pass]

--details = detailed view, usually not required (must placed BEFORE everything else)


********************************************************************

v0.5

Works only for non-parallel tasks.

Usage:
job_memusage.exe -r filename-to-run [-a arguments-to-pass] [-d]

(-d = detailed view, usually not required)

********************************************************************


log_memusage.h
#ifndef LOG_MEMUSAGE_H
#define LOG_MEMUSAGE_H

#ifdef __cplusplus
extern "C" {
#endif

  /**
   * \mainpage
   * \p log \p memusage is a shared library and optional API that monitors an applications memory usage.
   */

  /**
     \page peak_memusage
     \p peak_memusage - Determine the high-water mark memory usage of an application.

     \section peak_memusageSYNOPSIS SYNOPSIS
     \verbatim
     $ peak_memusage <program> --args --for-program

     $ mpiexec peak_memusage <mpi_program>
     \endverbatim

     \section peak_memusageDESCRIPTION DESCRIPTION
     peak_memusage determines the high-water mark of an
     application's memory usage.

     This utility is a wrapper around \p getrusage, which reports peak
     memory use of any executable. It is OpenMP and MPI aware and
     tries to report thread- and task- specific data. Of course,
     being OpenMP shared memory, that report cannot really separate
     thread-specific memory use.

     Note that the memory usage is reported only at the end of the
     run.

     \section peak_memusageOPTIONS OPTIONS
     The peak_memusage does not take any options.

     \section peak_memusageEXAMPLES EXAMPLES
     \verbatim
     $ ./peak_memusage ./use_memory
     Running: ./use_memory; sleep 1  - Please wait...
     Allocating, using and freeing 1000000 ints (3.81 MiB) in thread 0/1
     casper-login2 used memory in task 0: 30.82MiB (+23.03MiB overhead). ExitStatus: 0. Signal: 0
     \endverbatim
  */

  /**
     \page log_memusage log_memusage
     \p log_memusage - Logs memory usage history of an application.

     \section log_memusageSYNOPSIS SYNOPSIS
     \verbatim
     $ gcc -o my_app my_app.c -llog_memusage
     $ ./my_app

     $ mpiexec -x LD_PRELOAD=/path/to/liblog_memusage.so <mpi_program>
     \endverbatim

     \section log_memusageDESCRIPTION DESCRIPTION
     \p log_memusage is a library that monitors and reports application memory usage.

     \section log_memusageENVIRONMENT ENVIRONMENT

     General Environment Variables

     \subsection  LOG_MEMUSAGE_VERBOSE
     If enabled (1), prints key library events to \p stderr.

     Default: 0

     \subsection  LOG_MEMUSAGE_ENABLE_LOGFILE
     If enabled (1), logs memory usage vs. time to the output file "memory_usage.log",
     or \p LOG_MEMUSAGE_LOGFILE_NAME.

     Default: 0

     \subsection  LOG_MEMUSAGE_LOGFILE_NAME
     When \p LOG_MEMUSAGE_ENABLE_LOGFILE is enabled, \p LOG_MEMUSAGE_LOGFILE_NAME specifies the name of the
     output logfile.  Can be a relative or full path.  MPI rank information will be appended, if detected at run time.

     Default: "memory_usage.log"

     \subsection  LOG_MEMUSAGE_POLL_INTERVAL
     Floating point value specifying the polling frequency (in seconds).

     Default: 0.1 (sec)

     \subsection  LOG_MEMUSAGE_OUTPUT_INTERVAL
     Floating point value specifying the output frequency (in seconds).  Must be greater than or equal to \p LOG_MEMUSAGE_POLL_INTERVAL.

     Default: 1.0 (sec)

     \subsection LOG_MEMUSAGE_TRIPWIRES LOG_MEMUSAGE_CPU_MEM_TRIPWIRE, LOG_MEMUSAGE_GPU_MEM_TRIPWIRE
     If enabled, kill the application immediately if it exceeds the specified value (MB).

     Default: \p INT_MAX (MB)

     \section log_memusageEXAMPLES EXAMPLES
     \subsection baseline Baseline uninstrumented program:

     \verbatim
     $ ./use_memory -s 50000000
     Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/1
     \endverbatim

     \subsection ld_preload Using LD_PRELOAD to instrument an existing program:
     \verbatim
     $ LD_PRELOAD=./liblog_memusage.so ./use_memory -s 50000000
     logging memory usage for process 103284 to memory_usage.log ...
     Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/1
     # (memusage) --> casper-login2 / PID 103284, peak used memory: 213 MiB
     \endverbatim

     \subsection ld_preload_openmp Using LD_PRELOAD to instrument an existing OpenMP program:
     \verbatim
     $ LD_PRELOAD=./liblog_memusage.so OMP_NUM_THREADS=6 ./use_memory_openmp -s 50000000
     logging memory usage for process 106876 to memory_usage.log ...
     Allocating, using and freeing 100000000 ints (381.47 MiB) in thread 1/6
     Allocating, using and freeing 200000000 ints (762.94 MiB) in thread 3/6
     Allocating, using and freeing 150000000 ints (572.20 MiB) in thread 2/6
     Allocating, using and freeing 50000000 ints (190.73 MiB) in thread 0/6
     Allocating, using and freeing 250000000 ints (953.67 MiB) in thread 4/6
     Allocating, using and freeing 300000000 ints (1144.41 MiB) in thread 5/6
     # (memusage) --> casper-login2 / PID 106876, peak used memory: 4036 MiB
     \endverbatim

     \subsection pkgconfig Using pkg-config configuration to link with an application:
     \verbatim
     # compile:
     $ gcc -o use_memory ./use_memory.c $(pkg-config log_memusage --libs --cflags)

     # run:
     $ ./use_memory
     logging memory usage for process 111615 to memory_usage.log ...
     Allocating, using and freeing 1000000 ints (3.81 MiB) in thread 0/1
     # (memusage) --> casper-login2 / PID 111615, peak used memory: 24 MiB
     \endverbatim
  */

  /**
   * Insert the specified \p annotation into the memory log file.
   * This is useful for example to annotate what phase of the main program is executing at a given time.
   */
  int  log_memusage_annotate (const char* annotation);


  /**
   * Gets the amount of CPU memory used.
   * \returns The amount of CPU memory currently in use (MB).
   */
  int  log_memusage_get ();


  /**
   * Prints the current memory usage status to \p stderr.
   * \returns The amount of CPU  memory currently in use (MB).
   */
  int  log_memusage_report (const char* prefix);


  /**
   * Pauses memory monitoring by termininaing the execution thread.
   */
  int  log_memusage_pause ();


  /**
   * Resumes memory monitoring by restarting the execution thread.
   */
  int  log_memusage_resume ();

  /**
   * Initializes the memory logging data structures and execution thread.
   */
  void log_memusage_initialize ();

  /**
   * Terminates the memory logging execution thread.
   */
  void log_memusage_finalize ();

  /**
   * Maximum number of GPUs supported on a single host.
   */
#ifndef LOG_MEMUSAGE_MAX_GPU_DEVICES
#  define LOG_MEMUSAGE_MAX_GPU_DEVICES 8
#endif

  /**
   * \p getrusage() returns bytes on OSX, kbytes on Linux.  Handle apropriately.
   */
#if defined(__APPLE__)
#   define LOG_MEMUSAGE_OS_RSUSAGE_TO_MB (1024*1024)
#else
#   define LOG_MEMUSAGE_OS_RSUSAGE_TO_MB (1024)
#endif

  /**
   * Structure for reporting memory usage for (potentially) multiple GPUs.
   */
  struct log_memusage_gpu_memory
  {
    /**
     * The number of GPU deviced dectected.
     */
    unsigned int device_count;

    /**
     * Current memory usage of each GPU (MB).
     */
    int used[LOG_MEMUSAGE_MAX_GPU_DEVICES];

    /**
     * Current free memory remaining of each GPU (MB).
     */
    int free[LOG_MEMUSAGE_MAX_GPU_DEVICES];

    /**
     * Maximum memory used on any of the \p device_count GPUs (MB).
     */
    int max_used;

    /**
     * Total memory used on all of the \p device_count GPUs (MB).
     */
    int total_used;
  };

  /**
   * Simple typedef for using \p log_memusage_gpu_memory structs.
   */
  typedef struct log_memusage_gpu_memory log_memusage_gpu_memory_t;

  /**
   * \returns the number of GPUs active for this \p main() application.
   */
  int log_memusage_ngpus ();

  /**
   * \returns the latest memory information for all GPUs.
   */
  log_memusage_gpu_memory_t log_memusage_get_each_gpu ();

  /**
   * \returns the total memory used on all of the \p device_count GPUs (MB).
   */
  int log_memusage_get_all_gpus ();

  /**
   * \returns the maximum memory used on any of the \p device_count GPUs (MB).
   */
  int log_memusage_get_max_gpu ();

#ifdef __cplusplus
}
#endif
#endif /* #define LOG_MEMUSAGE_H */

//  LocalWords:  memusage subsubsection LOGFILE logfile
